src/main/java/org/apache/sling/scripting/core/impl/bundled/AbstractBundledRenderUnit.java

```diff
@@ -101,83 +95,37 @@ public BundleContext getBundleContext() {
         return scriptExtension;
     }
 
+    @Override
+    public @NotNull ServiceCache getServiceCache() {
+        return serviceCache;
+    }
+
     @Override
     @Nullable
     @SuppressWarnings("unchecked")
     public <T> T getService(@NotNull String className) {
-        LOG.debug("Attempting to load class {} as an OSGi service.", className);
-        T result = (this.services == null ? null : (T) this.services.get(className));
-        if (result == null) {
-            final ServiceReference<?> ref = this.bundleContext.getServiceReference(className);
-            if (ref != null) {
-                result = (T) this.bundleContext.getService(ref);
-                if (result != null) {
-                    if (this.services == null) {
-                        this.services = new HashMap<>();
-                    }
-                    if (this.references == null) {
-                        this.references = new ArrayList<>();
-                    }
-                    this.references.add(ref);
-                    this.services.put(className, result);
-                    return result;
-                }
-            }
+        try {
+            ClassLoader bundleClassloader = getBundle().adapt(BundleWiring.class).getClassLoader();
+            return (T) serviceCache.getService(bundleClassloader.loadClass(className));
+        } catch (ClassNotFoundException e) {
```

Comment:
I guess this is quite a change so - previously, we would have allowed to lookup up services with interfaces that are not wired to the bundle. Now, that will fail - are we sure that nobody is using this to lookup services that are then used only via reflection (i.e., don't need the interfaces to be available)?

Reply:
In HTL we always made sure that the object to load was accessible to the classloader with which the script was loaded. This was not necessarily the case for JSPs, but for bundled scripts they worked like that from the beginning - see https://github.com/apache/sling-org-apache-sling-scripting-spi/blob/47949fedc62380822961d8ef649130dee68a28e8/src/main/java/org/apache/sling/scripting/spi/bundle/BundledRenderUnit.java#L83.